lib/lower/Bridge.cpp

```diff
@@ -649,36 +649,42 @@ class FirConverter : public AbstractConverter {
 
   void genCharacterAssignement(
       const Ev::Assignment::IntrinsicAssignment &assignment) {
+    // TODO: Move these helpers in a way they can be used here,
+    // in ConvertExpr and IO.
+
+    // Helper to get length from expression
+    auto getCharLen{[&](const SomeExpr &charExpr) {
+      const auto &someCharExpr{
+          std::get<Ev::Expr<Ev::SomeCharacter>>(charExpr.u)};
+      /// FIXME: Not sure length can always be computed with LEN().
+      std::optional<Ev::Expr<Ev::SubscriptInteger>> lenExpr{someCharExpr.LEN()};
```

Comment:
Yeah. It seems to work for a literal, but a general CHARACTER variable could have a deferred length, so...

Reply:
It actually also works for stuff that are deferred length (the expression is then a `DesrciptorInquiry`), but it does not work for tough intrinsics (like `TRIM` where the result length basically nearly require running the TRIM algorithm) or user function call that returns character.

This will become useless in my proposed next change where
`Epxr<SomeCharacter>` are lowered to one of the following mlir::value:
-  fir.char<kind> :  length one (length know at compile time) characters.
- `fir.box_char<>` containing a `fir.ref<fir.array<len x shape x fir.char<kind>>>`
-  A `fir.box` (for non simply contiguous stuff).
The the length is either 1, or in the `fir.box_char`/`fir.box`.
That will simplify a lot the way we interact with character expressions.

Reply:
Ah, ok. I see what you mean in the comment now. Thanks.

So:
A single character has type `!fir.char<k>`, where k is a "kind" constant.

A sequence of characters has type `!fir.array<n x !fir.char<k>>`, where n is a "len" constant.

A simple argument descriptor for passing a CHARACTER variable is `!fir.boxchar<k>`. The descriptor must be built with `fir.embox_char` with a memory reference SSA-value of type `!fir.ref<!fir.char<k>>` and an integer SSA-value for the LEN type parameter.

The generic argument descriptor can be used to pass anything. But in the case of passing a simple CHARACTER(LEN=n,KIND=k) variable it's `!fir.box<!fir.array<n x !fir.char<k>>`.  That is built using `fir.embox` and would take a `!fir.ref<!fir.array<n x !fir.char<k>>>` and a 0 rank dimension.